### PR TITLE
mtmd: server: Support multimodal data prompt in /completions and /embeddings endpoint of server

### PR DESCRIPTION
**Pull request by 65a**
Editing first comment to match current state:

This pull adds support for a multimodal data in the `/completions` (and in a similar fashion, `/embeddings`) API endpoint. Instead of a string, list of tokens, or a mixed string token list as currently supported by that endpoint, this pull add support for a JSON object containing both a `prompt_string` and and a `multimodal_data` field. The client should check the result of `/models` or `/v1/models` for the `multimodal` capability, sending multimodal data to a non-multimodal model will result in a request error.

A singular request example is like this:
```
 {
    ...
    "prompt": { "prompt_string": "string", "multimodal_data": ["base64"] }
    ...
{
```

A multiple (and mixed-type) request would look like:
```
 {
    ...
    "prompt": [
        { "prompt_string": "string", "multimodal_data": ["base64"] },
        "What's up?",
        [ 0, 1, 2, 3, "breakfast" ]
    ]
    ...
{
```
All existing tests pass, new tests are added to cover both the prompt splitting and visual inference. If a prompt is added and the model does not support MTMD, an error is returned. Checking for the `multimodal` in capability `/models` API serves two purposes: it means the model supports it and the llama.cpp is new enough to support this interface. The multimodal part should be a base64-encoded media data supported by libmtmd.

With this approach, other server endpoints can be multimodal in the future relatively easily (Rerank and infill are close but would need additional work and testing, etc). Feedback welcome!

~~Implement a basic way to include multimodal data in the completions endpoint. For now, this just supports directly included data, in base64-encoded format provided as an array of strings under the json key `multimodal_data`. Documentation updated to match. Local testing shows no regression for without-media case, and successful image processing with media provided from a custom client.~~

~~Similar to #14016 but avoids importing the URL fetching logic at all. It could be added later when factored out of the OpenAI-emulation code, but this is simpler for now and avoids need for URL parsing and remote fetch capabilities.~~

Original referenced issue was #13872 

@ngxson ptal, hopefully this works for you.

### CODE REVIEW

File: tools/server/README.md [hunk 1]
```diff
@@ -407,6 +408,8 @@ Multiple prompts are also supported. In this case, the completion result will be
   - Strings and sequences of tokens: `["string1", [12, 34, 56]]`
   - Mixed types: `[[12, 34, "string", 56, 78], [12, 34, 56], "string"]`
 
+`multimodal_data`: Optional. Should be an array of strings, containing base64 encoded multimodal data (e.g. images, audio). There must be an identical number of MTMD media markers in the raw prompt which act as placeholders for the data provided to this parameter. The multimodal data files will be substituted in order.
```
Comment (pnb) [inline tools/server/README.md]: Is it correct that the "media markers" mentioned here should be `<__media__>`? I could not figure this out from the docs so it might be good to clarify for users. I found `<__media__>` here:
https://github.com/ggml-org/llama.cpp/blob/476aa3fd5779b32d06cd84338f777da82341195c/tools/mtmd/mtmd.h#L178

Comment (65a) [reply tools/server/README.md]: It is _currently_ `<__media__>`, and was `<__image__>`, but it appears the correct thing to do is use the marker returned by `mtmd_default_marker()`. I assume this means the marker could continue to change.

Comment (65a) [reply tools/server/README.md]: Done. Added some text clarifying the markers and linking to the mtmd C API header where the function is defined, but also provided the current value as an example.

**Comment by ngxson** [comment]
The proposal looks ok but there will be some edge cases:
- What if user enter a list of tokens or mixed list of tokens instead of a raw text prompt?
- What if user enter a list of multiple prompts?

I think proper test cases are required for this PR, similar to `/chat/completion`

**Comment by 65a** [comment]
Let me look at those.

The token case is interesting, interested in your thoughts there. The current server uses null tokens, but already knows how many to insert, which seems hard on the client (have to know the multimodal embedding size before sending raw tokens to completion endpoint). A magic token could work similarly to `<__media__>` but I guess this might require behavior changes. Need to poke at code to understand that path better, `test_completions.py` has no tests for token prompt requests. Easy option is to just document token + multimodal is not yet supported and throw if it happens.

The multiple text prompt part is also interesting from a usability perspective. I'll think about these and come back. The multi-prompt case should be straightforward to add tests for, not sure how it ought to work yet.

**Comment by 65a** [comment]
I have an idea that might be usable, namely that prompt can now contain an array of JSON objects, like  `"prompt": [{ prompt: "foo", multimodal_data: ["<base64>"] }, { prompt: "bar", multimodal_data: ["<base64>"] }]`. This would be a specific documented option for the prompt field in /completions, and only allow string multimodal prompts (for now). It's a bit verbose and requires more code refactoring, but as a user it makes sense to pair my prompts with their respective data. This would extend the currently supported prompt field options, and not change the top level data JSON object. Also easy to add additional tests for. I'll try it out locally, but does that idea work for you @ngxson ?

**Comment by 65a** [comment]
Rough draft for the idea here (it compiles and passes existing tests): https://pastebin.com/8zek7ium

Not complete or properly indented, but the idea is to use server_tokens in more places, so that the input tokenizer can branch and use MTMD tokenization where it makes sense to do so. As a side effect, probably got multimodal support in embeddings. Infill needs more work, and rerank would work if I can get the push_back(server_tokens) for server_tokens to work properly, I think.

There are probably better ways to do some of this than I did, feedback welcome.

**Comment by 65a** [comment]
Improved version of the rough draft that actually works, ignore indentation: https://pastebin.com/R6NdKQPP

This works locally for my use case, and I've started adding tests. There are a few TODOs to make doc ranking and embeddings support multimodal usecases, and I think the oai case can also be streamlined.

The general approach is as described previously: use server_tokens in more places, break out mtmd prompt parsing into a function, and change various input tokenization calls to handle server_tokens instead of llama_tokens.

The request format for multiple prompts would be like this:
```
{
    ...
    "prompt": [
            "Prompt 1",
            [ 1, 2, 3 , 4],
            { "prompt": "What is a tomato?", "multimodal_data": "<base64>"
    ],
    ....
}
```
The JSON entry only supports what `mtmd_tokenize` supports if multimodal_data is present. A single JSON object can also be provided instead of the array of prompts as above. A JSON object could also be provided containing only the prompt in either location, which is handled similarly to normal prompt strings.

**Comment by 65a** [comment]
Added tests including vision test. Should be good for a review pass. There is some potential future work, including supporting multimodal prompts in document rerank and infill. Embeddings may already work, existing tests pass, but I didn't try it and not sure it's expected to provide a stable embedding or not. Further refactoring is possible to streamline the OAI chat path into the rest, but probably a follow up. @ngxson let me know what you think.

**Comment by 65a** [comment]
Cleaned up the code quite a bit, and fixed the TODO around server_tokens.push_back(server_tokens). Now the tokenize_inputs handling reads a lot cleaner, which is nice.

**Comment by oobabooga** [comment]
I have tested this PR and it worked perfectly ✅

Here is a simple test with `google_gemma-3-4b-it-Q4_K_S.gguf`:

print

The prompt was:

```
<bos><start_of_turn>user
<__media__>

What do you see here?<end_of_turn>
<start_of_turn>model
```

The details of my UI integration are here https://github.com/oobabooga/text-generation-webui/pull/7027

File: tools/server/tests/unit/test_vision_completion.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+@pytest.mark.parametrize(
+    "prompt, image_data, success, re_content",
+    [
+        # test model is trained on CIFAR-10, but it's quite dumb due to small size
+        ("What is this: <__media__>\n", IMG_BASE64_0,           True, "(cat)+"), # exceptional, so that we don't cog up the log
+        ("What is this: <__media__>\n", IMG_BASE64_1,           True, "(frog)+"),
+        ("What is this: <__media__>\n", "malformed",              False, None),
+        ("What is this:\n", "base64",        False, None), # non-image data
```
Comment (pnb) [inline tools/server/tests/unit/test_vision_completion.py]: Not that it truly matters but I don't see much consistency in the whitespace for these few lines above, whereas the rest of the code seems more intentional.

Comment (65a) [reply tools/server/tests/unit/test_vision_completion.py]: Fixed.

File: tools/server/tests/unit/test_vision_completion.py
Comment (ngxson) [inline tools/server/tests/unit/test_vision_completion.py:1]: Can we merge this file to `test_vision_api.py`? We don't have many tests atm, so we should reduce number of files

File: tools/server/utils.hpp [hunk 3]
```diff
@@ -1356,3 +1316,132 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
...
+ * - "prompt": ["string1", [12, 34, 56]]
+ * - "prompt": [[12, 34, 56], [78, 90, 12]]
+ * - "prompt": [[12, 34, "string", 56, 78], [12, 34, 56], { "prompt": "string", "multimodal_data": [ "base64" ]}]
+ */
+static std::vector<server_tokens> tokenize_input_prompts(const llama_vocab * vocab, mtmd_context * mctx, const json & json_prompt, bool add_special, bool parse_special) {
+    std::vector<server_tokens> result;
+    if (json_prompt.is_array() && !json_is_array_with_tokens(json_prompt)) {
+        result.reserve(json_prompt.size());
+        for (const auto & p : json_prompt) {
+            result.push_back(tokenize_input_subprompt(vocab,mctx, p,add_special, parse_special));
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
            result.push_back(tokenize_input_subprompt(vocab,mctx, p, add_special, parse_special));
```

File: tools/server/utils.hpp [hunk 4]
```diff
@@ -1356,3 +1316,132 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
...
+    const bool has_mtmd = mctx != nullptr;
+    if (json_prompt.is_string() || json_is_array_of_mixed_numbers_strings(json_prompt)) {
+        // string or mixed
+        llama_tokens tmp = tokenize_mixed(vocab, json_prompt, add_special, parse_special);
+        return server_tokens(tmp, false);
+    } else if (json_is_array_of_numbers(json_prompt)) {
+        // array of tokens
+        llama_tokens tmp = json_prompt.get<llama_tokens>();
+        return server_tokens(tmp, false);
+    } else if (json_prompt.find("prompt") != json_prompt.end()) {
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
    } else if (json_prompt.contains("prompt")) {
```

File: tools/server/utils.hpp [hunk 5]
```diff
@@ -1356,3 +1316,132 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
...
+        // JSON object with prompt key.
+        if (has_mtmd && json_prompt.find("multimodal_data") != json_prompt.end()) {
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
        if (has_mtmd && json_prompt.contains("multimodal_data")) {
```

Or even better, like this:

```
if (json_prompt.contains("multimodal_data")) {
  if (has_mtmd) { ... do the thing .... }
  else throw std::runtime_error("multimodal is not supported by this server");
}
```

File: tools/server/server.cpp [hunk 6]
```diff
@@ -4750,22 +4709,22 @@ int main(int argc, char ** argv) {
             return;
         }
 
-        llama_tokens tokenized_query = tokenize_input_prompts(ctx_server.vocab, query, /* add_special */ false, true)[0];
+        server_tokens tokenized_query = std::move(tokenize_input_prompts(ctx_server.vocab, ctx_server.mctx, query, /* add_special */ false, true)[0]);
```
Comment (ngxson) [inline tools/server/server.cpp]: I think `std::move` is unnecessary here, the compiler should be good enough to optimize this

**Review by ngxson — COMMENTED** [review]
Looking good, can be merge after my comments are all resolved

File: tools/server/tests/unit/test_vision_completion.py
Comment (65a) [reply tools/server/tests/unit/test_vision_completion.py:1]: Done. Added some tests for other functionality discussed below, and updated test_completions.py (it doesn't actually have MTMD, so dropped the data).

File: tools/server/utils.hpp [hunk 7]
```diff
@@ -28,6 +28,8 @@
 #include <cinttypes>
 
 #define DEFAULT_OAICOMPAT_MODEL "gpt-3.5-turbo"
+#define JSON_STRING_PROMPT_KEY "prompt_string"
+#define JSON_MTMD_DATA_KEY "multimodal_data"
```
Comment (ngxson) [inline tools/server/utils.hpp]: it's best to define these in local scope, not globally, using `const char *`

File: tools/server/utils.hpp (hunk 3 above)
Comment (65a) [reply tools/server/utils.hpp]: Done.

File: tools/server/utils.hpp (hunk 4 above)
Comment (65a) [reply tools/server/utils.hpp]: Done.

File: tools/server/utils.hpp [hunk 8]
```diff
@@ -1356,3 +1318,135 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
...
+        return server_tokens(tmp, false);
+    } else if (json_prompt.contains(JSON_STRING_PROMPT_KEY)) {
+        // JSON object with prompt key.
+        if (json_prompt.contains(JSON_MTMD_DATA_KEY)) {
+            if (!has_mtmd)
+                throw std::runtime_error("Multimodal data provided, but model does not support multimodal requests.");
+
+            // JSON object with prompt and multimodal key.
+            std::vector<raw_buffer> files;
+            for (const auto& entry : json_prompt.at(JSON_MTMD_DATA_KEY)) {
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
            for (const auto & entry : json_prompt.at(JSON_MTMD_DATA_KEY)) {
```

File: tools/server/utils.hpp (hunk 5 above)
Comment (65a) [reply tools/server/utils.hpp]: Done, however this leaves a trap for the clients who call us, because they don't know if we will give them an error or not. Added a capability to /models and /v1/models so the client can tell if we support multimodal or not, this should be sufficient to support an error rather than silently dropping it.

File: tools/server/utils.hpp [hunk 9]
```diff
@@ -1356,3 +1318,135 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
...
+    auto bitmaps_c_ptr = bitmaps.c_ptr();
+    int32_t tokenized = mtmd_tokenize(mctx,
+                                      chunks.ptr.get(),
+                                      &inp_txt,
+                                      bitmaps_c_ptr.data(),
+                                      bitmaps_c_ptr.size());
+    if (tokenized != 0) {
+        throw std::runtime_error("Failed to tokenize prompt");
+    }
+    auto result = server_tokens(chunks,true);
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
    auto result = server_tokens(chunks, true);
```

File: tools/server/server.cpp (hunk 6 above)
Comment (65a) [reply tools/server/server.cpp]: Done. It was necessary, at least on my recent gcc on linux (this was triggering an ateempted copy), but I refactored this so we only std::move once, below this point.

File: tools/server/tests/utils.py [hunk 10]
```diff
@@ -478,10 +478,11 @@ def tinygemma3() -> ServerProcess:
         server.mmproj_url = "https://huggingface.co/ggml-org/tinygemma3-GGUF/resolve/main/mmproj-tinygemma3.gguf"
         server.model_alias = "tinygemma3"
         server.n_ctx = 1024
-        server.n_batch = 32
+        server.n_batch = 512
         server.n_slots = 2
         server.n_predict = 4
         server.seed = 42
+        server.server_embeddings = True
         return server
```
Comment (ngxson) [inline tools/server/tests/utils.py]: define this in local test, before `server.start(...)`

see other test files for example

**Comment by 65a** [comment]
A few notable updates:
1. Refactored `prompt` and `multimodal_data` to `prompt_string` and `multimodal_data`. This avoids having duplicate keys and opens up the possibility of having things like `prompt_tokens` or `multimodal_urls` in the future.
2. Sending multimodal data to a model without multimodal capability now results in a request error. To avoid that, clients should check `/v1/models` or `/models` for the `multimodal` capabability. Tests added for this new capability as well.
3. Embeddings work and are now documented and tested as multimodal
4. Removed some text about `image_data` in the docs that didn't appear to be implemented yet, more docs for multimodal
5. Made test names better and merged them into the one file.

@ngxson ready for re-review, I won't resolve your comments in case you want to discuss any of the changes.

File: tools/server/utils.hpp (hunk 7 above)
Comment (65a) [reply tools/server/utils.hpp]: Done

File: tools/server/utils.hpp (hunk 8 above)
Comment (65a) [reply tools/server/utils.hpp]: Done.

File: tools/server/utils.hpp (hunk 9 above)
Comment (65a) [reply tools/server/utils.hpp]: Done.

File: tools/server/tests/utils.py (hunk 10 above)
Comment (65a) [reply tools/server/tests/utils.py]: Done.

**Comment by 65a** [comment]
@ngxson I believe second round of comments are now addressed!

@oobabooga I saw you already were testing this, thanks. Please note the API has changed slightly. The client should check if multimodal is supported via the `/models` endpoint before sending a prompt that includes the `multimodal_data` key. The `prompt` key was renamed to `prompt_string`. Thanks for testing it!

**Comment by oobabooga** [comment]
Thanks for the heads up @65a, I have updated the request! https://github.com/oobabooga/text-generation-webui/commit/e6447cd24acbde845dbb4aa27acfd4c17b5c849c

**Comment by 65a** [comment]
Looks like the UNDEFINED sanitizer is broken at HEAD by an unrelated change somewhere in ggml :cry:

**Comment by 65a** [comment]
Can't reproduce sanitizer test timeout failure locally, hopefully pushing an updated commit message can retrigger CI.

**Comment by 65a** [comment]
@ngxson resolved trivial fixes, left the more interesting ones open to see if you're happy. Thanks for your review.

File: tools/server/utils.hpp (hunk 5 above)
Comment (65a) [reply tools/server/utils.hpp]: Tested this approach locally. It works well for me, because then if the llama.cpp server is old, it will never have the mutlimodal capability, and so I will never accidentally send a multimodal request to either an old server or a model that doesn't support it.

**Comment by 65a** [comment]
Rebased on HEAD, resolving open comments. Please feel free to re-open or comment if there's anything remaining @ngxson

**Comment by oobabooga** [comment]
I have merged this PR in [my fork of llama.cpp](https://github.com/oobabooga/llama.cpp) and am distributing binaries including the changes here. Many people have already tested it and verified the implementation works.

**Comment by ngxson** [comment]
I was not available the past few days. Will review this PR tomorrow.

File: tools/server/tests/unit/test_vision_api.py [hunk 11]
```diff
@@ -10,21 +10,48 @@
...
+JSON_PROMPT_STRING_KEY = "prompt_string"
 
 @pytest.fixture(autouse=True)
 def create_server():
     global server
     server = ServerPreset.tinygemma3()
 
+def test_models_supports_multimodal_capability():
+    global server
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
Comment (ngxson) [inline tools/server/tests/unit/test_vision_api.py]:
```suggestion
    server.start()
```

the model is not big, just 50MB. it should be a problem on your side

File: tools/server/tests/unit/test_vision_api.py [hunk 12]
```diff
@@ -10,21 +10,48 @@
...
+    res = server.make_request("GET", "/models", data={})
+    assert res.status_code == 200
+    model_info = res.body["models"][0]
+    print(model_info)
+    assert "completion" in model_info["capabilities"]
+    assert "multimodal" in model_info["capabilities"]
+
+def test_v1_models_supports_multimodal_capability():
+    global server
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
Comment (ngxson) [inline tools/server/tests/unit/test_vision_api.py]:
```suggestion
    server.start()
```

File: tools/server/tests/unit/test_vision_api.py [hunk 13]
```diff
@@ -58,3 +85,62 @@ def test_vision_chat_completion(prompt, image_url, success, re_content):
...
+        # test model is trained on CIFAR-10, but it's quite dumb due to small size
+        ("What is this: <__media__>\n", IMG_BASE64_0,           True, "(cat)+"),
+        ("What is this: <__media__>\n", IMG_BASE64_1,           True, "(frog)+"),
+        ("What is this: <__media__>\n", "malformed",            False, None), # non-image data
+        ("What is this:\n",             "",                     False, None), # empty string
+    ]
+)
+def test_vision_completion(prompt, image_data, success, re_content):
+    global server
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
Comment (ngxson) [inline tools/server/tests/unit/test_vision_api.py]:
```suggestion
    server.start()
```

File: tools/server/tests/unit/test_vision_api.py [hunk 14]
```diff
@@ -58,3 +85,62 @@ def test_vision_chat_completion(prompt, image_url, success, re_content):
...
+        ("What is this: <__media__>\n", IMG_BASE64_1,           True),
+        ("What is this: <__media__>\n", "malformed",            False), # non-image data
+        ("What is this:\n",             "base64",               False), # non-image data
+    ]
+)
+def test_vision_embeddings(prompt, image_data, success):
+    global server
+    server.server_embeddings=True
+    server.n_batch=512
+    server.start(timeout_seconds=60) # vision model may take longer to load due to download size
```
Comment (ngxson) [inline tools/server/tests/unit/test_vision_api.py]:
```suggestion
    server.start()
```

File: tools/server/utils.hpp [hunk 15]
```diff
@@ -123,6 +123,19 @@ static bool json_is_array_of_mixed_numbers_strings(const json & data) {
     return false;
 }
 
+// does array have any individual integers/tokens?
+static bool json_is_array_with_tokens(const json & data) {
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
static bool json_is_array_contains_number(const json & data) {
```

File: tools/server/utils.hpp [hunk 16]
```diff
@@ -1356,3 +1316,137 @@ static std::string fnv_hash(const uint8_t * data, size_t len) {
...
+ * - "prompt": { "prompt_string": "string", "multimodal_data": [ "base64" ] }
+ * and multiple prompts (multi-tasks):
+ * - "prompt": ["string1", "string2"]
+ * - "prompt": ["string1", [12, 34, 56]]
+ * - "prompt": [[12, 34, 56], [78, 90, 12]]
+ * - "prompt": [[12, 34, "string", 56, 78], [12, 34, 56], { "prompt_string": "string", "multimodal_data": [ "base64" ]}]
+ */
+static std::vector<server_tokens> tokenize_input_prompts(const llama_vocab * vocab, mtmd_context * mctx, const json & json_prompt, bool add_special, bool parse_special) {
+    std::vector<server_tokens> result;
+    if (json_prompt.is_array() && !json_is_array_with_tokens(json_prompt)) {
```
Comment (ngxson) [inline tools/server/utils.hpp]:
```suggestion
    if (json_prompt.is_array() && !json_is_array_contains_number(json_prompt)) {
```

File: tools/server/tests/unit/test_vision_api.py (hunk 11 above)
Comment (65a) [reply tools/server/tests/unit/test_vision_api.py]: Done.

File: tools/server/tests/unit/test_vision_api.py (hunk 12 above)
Comment (65a) [reply tools/server/tests/unit/test_vision_api.py]: Done.

File: tools/server/tests/unit/test_vision_api.py (hunk 13 above)
Comment (65a) [reply tools/server/tests/unit/test_vision_api.py]: Done.

File: tools/server/tests/unit/test_vision_api.py (hunk 14 above)
Comment (65a) [reply tools/server/tests/unit/test_vision_api.py]: Done.

File: tools/server/utils.hpp (hunk 15 above)
Comment (65a) [reply tools/server/utils.hpp]: Done.

File: tools/server/utils.hpp (hunk 16 above)
Comment (65a) [reply tools/server/utils.hpp]: Done.

**Comment by 65a** [comment]
@ngxson comments fixed and resolved, rebased on HEAD.

**Comment by 65a** [comment]
Looks good now?

**Review by ngxson — APPROVED** [review]
